metricbeat/internal/metrics/cpu/metrics_linux.go

```diff
@@ -0,0 +1,92 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cpu
+
+import (
+	"bufio"
+	"strings"
+
+	"github.com/joeshaw/multierror"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/metricbeat/internal/metrics"
+)
+
+func scanStatFile(scanner *bufio.Scanner) (CPUMetrics, error) {
+	cpuData, err := statScanner(scanner, parseCPULine)
+	if err != nil {
+		return CPUMetrics{}, errors.Wrap(err, "error scanning stat file")
+	}
+	return cpuData, nil
+}
+
+func parseCPULine(line string) (CPU, error) {
+	cpuData := CPU{}
+	fields := strings.Fields(line)
+	var errs multierror.Errors
+
+	user, err := touint(fields[1])
+	if err != nil {
+		errs = append(errs, err)
+	}
+	cpuData.User = metrics.OptUintWith(user)
```

Comment:
We never return early. As we have an OptUint, do we really want to report 0, or shall the optional be unset?


Not really required to change this here, but ways to reduce repetition:

```
tryParseUint := func(name, field string) (v metrics.OptUint) {
  u, err := touint(field)
  if err != nil {
    errs = append(errs, fmt.Errorf("failed to parse %v: %w", name, field))
  } else {
    v = metrics.OptUintWith(u)
  }
  return v
}
```

and use it like:

```
cpuData.User = tryParseUint("user", fields[1])
cpuData.Nice = tryparseUint("nice", fields[2])
```

Now the error message tells you which field failed to parse, plus we don't have to live with all those if-else blocks.
The tryParseUint not return an uninitialized Opt value if parsing failed. The curent code just returns 0. Not sure which one would fit better here. But  given we have an optional type I would assume we don't use 0.

For 'bigger' use cases one might want to turn the closer into a custom type.

In case you want to allow custom parser functions (hex vs. decimal representation) the helper could be written like this:


```
tryParse := func(u uint, err error) (v metrics.OptUint) {
  if err != nil {
    errs = append(errs, error)
  } else {
    v = metrics.OptUintWith(u)
  }
  return v
}
```

This function will be used like this (if number of returned values matches number of parameters we can just chain them):

```
cpuData.User = tryParse(touint(fields[1])
cpuData.Nice = tryPaese(touint(fields[2])
```

Unfortunately with that approach we loose the nice error messages.